From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Mon, 28 Jun 2021 12:52:10 +0300
Subject: [PATCH 04/67] Read rustc host from environment. JB#53019 OMP#JOLLA-93

---
 build/moz.configure/rust.configure | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/build/moz.configure/rust.configure b/build/moz.configure/rust.configure
index a89585f6f81d..2e6921ab719f 100644
--- a/build/moz.configure/rust.configure
+++ b/build/moz.configure/rust.configure
@@ -99,6 +99,7 @@ set_config("RUSTC", rustc)
 
 @depends_if(rustc)
 @checking("rustc version", lambda info: info.version)
+@imports("os")
 def rustc_info(rustc):
     if not rustc:
         return
@@ -107,7 +108,7 @@ def rustc_info(rustc):
     return namespace(
         version=Version(info.get("release", "0")),
         commit=info.get("commit-hash", "unknown"),
-        host=info["host"],
+        host=os.environ.get("RUST_HOST_TARGET", info["host"]),
         llvm_version=Version(info.get("LLVM version", "0")),
     )
 
