From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Thu, 8 Dec 2016 14:32:54 +0200
Subject: [PATCH] [sailfishos][compositor] Fix GLContextProvider defines

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
Signed-off-by: Pavel Tumakaev <p.tumakaev@omprussia.ru>
---
 gfx/gl/GLContextProvider.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/gfx/gl/GLContextProvider.h b/gfx/gl/GLContextProvider.h
index 2fa98581ba6e..d2868f6169da 100644
--- a/gfx/gl/GLContextProvider.h
+++ b/gfx/gl/GLContextProvider.h
@@ -48,12 +48,14 @@ namespace gl {
 #  define GL_CONTEXT_PROVIDER_DEFAULT GLContextProviderGLX
 #endif
 
+#ifndef MOZ_GL_PROVIDER
 #define GL_CONTEXT_PROVIDER_NAME GLContextProviderEGL
 #include "GLContextProviderImpl.h"
 #undef GL_CONTEXT_PROVIDER_NAME
 #ifndef GL_CONTEXT_PROVIDER_DEFAULT
 #  define GL_CONTEXT_PROVIDER_DEFAULT GLContextProviderEGL
 #endif
+#endif
 
 #if defined(MOZ_WAYLAND)
 #  define GL_CONTEXT_PROVIDER_NAME GLContextProviderWayland
@@ -76,6 +78,7 @@ namespace gl {
 #  define GL_CONTEXT_PROVIDER_NAME MOZ_GL_PROVIDER
 #  include "GLContextProviderImpl.h"
 #  undef GL_CONTEXT_PROVIDER_NAME
+#  undef GL_CONTEXT_PROVIDER_DEFAULT
 #  define GL_CONTEXT_PROVIDER_DEFAULT MOZ_GL_PROVIDER
 #endif
 
-- 
2.31.1

