From 0679595a70ff9f5f96c6f68f990b32fd17a905ff Tue, 7 Apr 2020 15:08:48 +0200
From: Andrew Branson <andrew.branson@jolla.com>
Date: Tue, 7 Apr 2020 00:02:45 +0200
Subject: [PATCH] [debug] Enable gstreamer pipeline debug graph generation. JB#49532

diff --git a/dom/media/gstreamer/GStreamerFunctionList.h b/dom/media/gstreamer/GStreamerFunctionList.h
index 75f352e..80d7819 100644
--- a/dom/media/gstreamer/GStreamerFunctionList.h
+++ b/dom/media/gstreamer/GStreamerFunctionList.h
@@ -35,6 +35,7 @@
 GST_FUNC(LIBGSTREAMER, gst_caps_new_full)
 GST_FUNC(LIBGSTREAMER, gst_caps_new_simple)
 GST_FUNC(LIBGSTREAMER, gst_caps_set_simple)
+GST_FUNC(LIBGSTREAMER, gst_debug_bin_to_dot_file_with_ts)
 GST_FUNC(LIBGSTREAMER, gst_element_factory_get_static_pad_templates)
 GST_FUNC(LIBGSTREAMER, gst_element_factory_get_type)
 GST_FUNC(LIBGSTREAMER, gst_element_factory_make)
diff --git a/dom/media/gstreamer/GStreamerReader.cpp b/dom/media/gstreamer/GStreamerReader.cpp
index 202f5a4..4183389 100644
--- a/dom/media/gstreamer/GStreamerReader.cpp
+++ b/dom/media/gstreamer/GStreamerReader.cpp
@@ -554,6 +554,9 @@
    * the appsinks */
   if (UpdateCurrentAsActiveIfNotBusy(this)) {
     gst_element_set_state(mPlayBin, GST_STATE_PLAYING);
+    GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS(GST_BIN(mPlayBin),
+                                       GstDebugGraphDetails(GST_DEBUG_GRAPH_SHOW_ALL /* GST_DEBUG_GRAPH_SHOW_MEDIA_TYPE | GST_DEBUG_GRAPH_SHOW_NON_DEFAULT_PARAMS | GST_DEBUG_GRAPH_SHOW_STATES*/),
+                                       "ReadMetadata");
   }
 
   return NS_OK;
