From c44992993cf5cad700d7c5836fa6de42f648f801 Mon Sep 17 00:00:00 2001
From: Andrew Branson <andrew.branson@jolla.com>
Date: Tue, 7 Apr 2020 11:12:19 +0200
Subject: [PATCH] Require interleaved audio layout. Fixes JB#49532

---
 dom/media/gstreamer/GStreamerReader.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dom/media/gstreamer/GStreamerReader.cpp b/dom/media/gstreamer/GStreamerReader.cpp
index 62dc770cd645..7f6836ee1834 100644
--- a/dom/media/gstreamer/GStreamerReader.cpp
+++ b/dom/media/gstreamer/GStreamerReader.cpp
@@ -1398,7 +1398,7 @@ void GStreamerReader::NotifyDataArrivedInternal()
 #if GST_VERSION_MAJOR >= 1
 GstCaps* GStreamerReader::BuildAudioSinkCaps()
 {
-  GstCaps* caps = gst_caps_from_string("audio/x-raw, channels={1,2}");
+  GstCaps* caps = gst_caps_from_string("audio/x-raw, channels={1,2}, layout=interleaved");
   const char* format;
 #ifdef MOZ_SAMPLE_TYPE_FLOAT32
 #if MOZ_LITTLE_ENDIAN
-- 
2.26.2

