From 4dc86eb20668f92f810388e07253b16bed1a152d Tue, 7 Apr 2020 15:08:37 +0200
From: Andrew Branson <andrew.branson@jolla.com>
Date: Tue, 7 Apr 2020 11:12:19 +0200
Subject: [PATCH] [media] Require interleaved audio layout. Fixes JB#49532

diff --git a/dom/media/gstreamer/GStreamerReader.cpp b/dom/media/gstreamer/GStreamerReader.cpp
index 4183389..b5cc71e 100644
--- a/dom/media/gstreamer/GStreamerReader.cpp
+++ b/dom/media/gstreamer/GStreamerReader.cpp
@@ -1340,7 +1340,7 @@
 #if GST_VERSION_MAJOR >= 1
 GstCaps* GStreamerReader::BuildAudioSinkCaps()
 {
-  GstCaps* caps = gst_caps_from_string("audio/x-raw, channels={1,2}");
+  GstCaps* caps = gst_caps_from_string("audio/x-raw, channels={1,2}, layout=interleaved");
   const char* format;
 #ifdef MOZ_SAMPLE_TYPE_FLOAT32
 #if MOZ_LITTLE_ENDIAN
