From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Llewellyn-Jones <david@flypig.co.uk>
Date: Thu, 2 Nov 2023 23:08:34 +0000
Subject: [PATCH 30/67] Make PresShell::SetIsActive() public

---
 layout/base/PresShell.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/layout/base/PresShell.h b/layout/base/PresShell.h
index 6223c638e306..6f6590a9ae4d 100644
--- a/layout/base/PresShell.h
+++ b/layout/base/PresShell.h
@@ -921,6 +921,7 @@ class PresShell final : public nsStubDocumentObserver,
     return mObservesMutationsForPrint;
   }
 
+  void SetIsActive(bool aIsActive) { SetIsActive(aIsActive, aIsActive); }
   void ActivenessMaybeChanged();
   // See ComputeActiveness() for details of these two booleans.
   bool IsActive() const { return mIsActive; }
-- 
2.51.0

