From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Llewellyn-Jones <david@flypig.co.uk>
Date: Wed, 5 Aug 2026 10:47:08 +0200
Subject: [PATCH 30/67] Make PresShell::SetIsActive() public

---
 layout/base/PresShell.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/layout/base/PresShell.h b/layout/base/PresShell.h
index 6f7443e16f420..283132846062a 100644
--- a/layout/base/PresShell.h
+++ b/layout/base/PresShell.h
@@ -928,6 +928,7 @@ class PresShell final : public nsStubDocumentObserver,
     return mObservesMutationsForPrint;
   }
 
+  void SetIsActive(bool aIsActive);
   void ActivenessMaybeChanged();
   bool IsActive() const { return mIsActive; }
 
@@ -1778,7 +1779,6 @@ class PresShell final : public nsStubDocumentObserver,
  private:
   ~PresShell();
 
-  void SetIsActive(bool aIsActive);
   bool ComputeActiveness() const;
 
   MOZ_CAN_RUN_SCRIPT
