From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pavel Tumakaev <p.tumakaev@omprussia.ru>
Date: Tue, 13 Apr 2021 13:37:00 +0300
Subject: [PATCH] [sailfishos][gecko] Disable link time optimization for rust
 components

---
 config/makefiles/rust.mk | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/config/makefiles/rust.mk b/config/makefiles/rust.mk
index 8a4782cc2c0f..530ae44e3d8b 100644
--- a/config/makefiles/rust.mk
+++ b/config/makefiles/rust.mk
@@ -55,18 +55,18 @@ endif
 # invocation (i.e., only the top-level crate, not its dependencies).
 cargo_rustc_flags = $(CARGO_RUSTCFLAGS)
 ifndef DEVELOPER_OPTIONS
-ifndef MOZ_DEBUG_RUST
+#ifndef MOZ_DEBUG_RUST
 # Enable link-time optimization for release builds, but not when linking
 # gkrust_gtest.
-ifeq (,$(findstring gkrust_gtest,$(RUST_LIBRARY_FILE)))
-cargo_rustc_flags += -Clto
-endif
+#ifeq (,$(findstring gkrust_gtest,$(RUST_LIBRARY_FILE)))
+#cargo_rustc_flags += -Clto
+#endif
 # Versions of rust >= 1.45 need -Cembed-bitcode=yes for all crates when
 # using -Clto.
-ifeq (,$(filter 1.38.% 1.39.% 1.40.% 1.41.% 1.42.% 1.43.% 1.44.%,$(RUSTC_VERSION)))
-RUSTFLAGS += -Cembed-bitcode=yes
-endif
-endif
+#ifeq (,$(filter 1.38.% 1.39.% 1.40.% 1.41.% 1.42.% 1.43.% 1.44.%,$(RUSTC_VERSION)))
+#RUSTFLAGS += -Cembed-bitcode=yes
+#endif
+#endif
 endif
 
 ifdef CARGO_INCREMENTAL
-- 
2.25.1

