commit 8fef9d3e8520a20eec8181ad95bcab6f1f68e813
Author: Dmitry Rozhkov <dmitry.rozhkov@jolla.com>
Date:   Wed Aug 6 12:55:50 2014 +0300

    Explicitly add headers for external types and values

diff --git a/embedding/embedlite/tests/gtest/TestEmbedLiteViewInit.cpp b/embedding/embedlite/tests/gtest/TestEmbedLiteViewInit.cpp
index c672caa..06dfc14 100644
--- a/embedding/embedlite/tests/gtest/TestEmbedLiteViewInit.cpp
+++ b/embedding/embedlite/tests/gtest/TestEmbedLiteViewInit.cpp
@@ -9,6 +9,7 @@
 #include "nsStringGlue.h"
 #include "nsIServiceManager.h"
 #include "nsIFile.h"
+#include "nsDirectoryServiceDefs.h" // for NS_GRE_DIR
 
 TEST(EmbedLiteViewInitTest, EmbedLiteAppCreateView)
 {
diff --git a/embedding/embedlite/utils/WebBrowserChrome.cpp b/embedding/embedlite/utils/WebBrowserChrome.cpp
index ea26941..92d8b3f 100644
--- a/embedding/embedlite/utils/WebBrowserChrome.cpp
+++ b/embedding/embedlite/utils/WebBrowserChrome.cpp
@@ -29,6 +29,7 @@
 #include "nsIURIFixup.h"
 #include "nsIEmbedBrowserChromeListener.h"
 #include "nsIBaseWindow.h"
+#include "ScriptSettings.h" // for AutoNoJSAPI
 
 #define MOZ_AFTER_PAINT_LITERAL "MozAfterPaint"
 #define MOZ_scroll "scroll"
