From 2d214b02cb06695a4c909d544007c8592e62fe82 Mon Sep 17 00:00:00 2001
From: Dmitry Rozhkov <dmitry.rozhkov@jolla.com>
Date: Thu, 18 Dec 2014 16:59:07 +0200
Subject: [PATCH] PATCH Disallow image locking no matter what

---
 content/base/src/nsDocument.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/content/base/src/nsDocument.cpp b/content/base/src/nsDocument.cpp
index 3493bce..8bd0003 100644
--- a/content/base/src/nsDocument.cpp
+++ b/content/base/src/nsDocument.cpp
@@ -10040,6 +10040,8 @@ nsDocument::SetImageLockingState(bool aLocked)
       !Preferences::GetBool("image.mem.allow_locking_in_content_processes", true)) {
     return NS_OK;
   }
+  // In embedlite we don't allow locking even if process type is not content.
+  return NS_OK;
 
   // If there's no change, there's nothing to do.
   if (mLockingImages == aLocked)
-- 
1.9.1

diff --git a/embedding/embedlite/embedding.js b/embedding/embedlite/embedding.js
index 67c5f75..571cbb2 100644
--- a/embedding/embedlite/embedding.js
+++ b/embedding/embedlite/embedding.js
@@ -261,7 +261,7 @@ pref("media.preload.auto", 2);    // preload metadata if preload=auto
 // optimize images memory usage
 pref("image.mem.decodeondraw", true);
 pref("image.mem.allow_locking_in_content_processes", false);
-pref("image.mem.min_discard_timeout_ms", 10000);
+pref("image.mem.min_discard_timeout_ms", 86400000); /* 24h, we rely on the out of memory hook */
 pref("image.onload.decode.limit", 24); /* don't decode more than 24 images eagerly */
 
 // SimplePush
