diff --git a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.cc b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.cc
index 7034eaa..4264b86 100644
--- a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.cc
+++ b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.cc
@@ -2360,12 +2360,14 @@ bool AudioDeviceLinuxALSA::KeyPressed() const{
   unsigned int i = 0;
   char state = 0;
 
+#ifdef MOZ_X11
   if (!_XDisplay)
     return false;
 
-#ifdef MOZ_X11
   // Check key map status
   XQueryKeymap(_XDisplay, szKey);
+#else
+  return false;
 #endif
 
   // A bit change in keymap means a key is pressed
diff --git a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.h b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.h
index dbe737e..ac20821 100644
--- a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.h
+++ b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_alsa_linux.h
@@ -19,8 +19,9 @@
 #include <alsa/asoundlib.h>
 #include <sys/soundcard.h>
 #include <sys/ioctl.h>
+#ifdef MOZ_X11
 #include <X11/Xlib.h>
-
+#endif
 
 namespace webrtc
 {
@@ -251,7 +252,9 @@ private:
     uint16_t _playBufDelayFixed;            // fixed playback delay
 
     char _oldKeyState[32];
+#ifdef MOZ_X11
     Display* _XDisplay;
+#endif
 };
 
 }
diff --git a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.cc b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.cc
index 2146319..f193862 100644
--- a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.cc
+++ b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.cc
@@ -3113,6 +3113,7 @@ bool AudioDeviceLinuxPulse::RecThreadProcess()
 
 bool AudioDeviceLinuxPulse::KeyPressed() const{
 
+#ifdef MOZ_X11
   char szKey[32];
   unsigned int i = 0;
   char state = 0;
@@ -3130,5 +3131,8 @@ bool AudioDeviceLinuxPulse::KeyPressed() const{
   // Save old state
   memcpy((char*)_oldKeyState, (char*)szKey, sizeof(_oldKeyState));
   return (state != 0);
+#else
+  return false;
+#endif
 }
 }
diff --git a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.h b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.h
index 3b8a181..951f46f 100644
--- a/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.h
+++ b/media/webrtc/trunk/webrtc/modules/audio_device/linux/audio_device_pulse_linux.h
@@ -16,7 +16,9 @@
 #include "critical_section_wrapper.h"
 
 #include <pulse/pulseaudio.h>
+#ifdef MOZ_X11
 #include <X11/Xlib.h>
+#endif
 
 // We define this flag if it's missing from our headers, because we want to be
 // able to compile against old headers but still use PA_STREAM_ADJUST_LATENCY
@@ -380,7 +382,9 @@ private:
     pa_buffer_attr _recBufferAttr;
 
     char _oldKeyState[32];
+#ifdef MOZ_X11
     Display* _XDisplay;
+#endif
 };
 
 }
