From 023a190bb938ce8750a53260a9eaca51c81e0b00 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jollamobile.com>
Date: Tue, 8 Apr 2014 18:43:40 +0300
Subject: [PATCH] Fix booster startup

---
 embedding/embedlite/utils/GeckoLoader.cpp | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/embedding/embedlite/utils/GeckoLoader.cpp b/embedding/embedlite/utils/GeckoLoader.cpp
index 159bcc3..a4330521 100644
--- a/embedding/embedlite/utils/GeckoLoader.cpp
+++ b/embedding/embedlite/utils/GeckoLoader.cpp
@@ -108,7 +108,7 @@ GeckoLoader::InitEmbedding(const char* aProfilePath)
   }
 
   // create nsIFile pointing to appdir
-  char self[MAX_PATH];
+  char self[MAX_PATH] = "";
 #ifdef WIN32
   GetModuleFileNameA(GetModuleHandle(NULL), self, sizeof(self));
 #else
@@ -122,11 +122,16 @@ GeckoLoader::InitEmbedding(const char* aProfilePath)
   size_t lastslash_t = selfPath.find_last_of("/\\");
   if (lastslash_t == std::string::npos) {
     LOGE("Invalid module filename: %s", self);
+#ifdef WIN32
     return false;
+#else
+    // A hackish way to get this initialized when starting through booster.
+    selfPath = std::string("/usr/bin");
+#endif
+  } else {
+    selfPath = selfPath.substr(0, lastslash_t);
   }
 
-  selfPath = selfPath.substr(0, lastslash_t);
-
   nsCOMPtr<nsIFile> appdir;
   rv = XRE_GetBinaryPath(selfPath.c_str(), getter_AddRefs(appdir));
   if (NS_FAILED(rv)) {
-- 
1.9.1

