diff --git a/layout/base/nsDisplayList.cpp b/layout/base/nsDisplayList.cpp
index bb03b11..dbadf55 100644
--- a/layout/base/nsDisplayList.cpp
+++ b/layout/base/nsDisplayList.cpp
@@ -720,7 +720,7 @@ static void RecordFrameMetrics(nsIFrame* aForFrame,
       ScreenIntRect screenBounds = ScreenIntRect::FromUnknownRect(mozilla::gfx::IntRect(
           bounds.x, bounds.y, bounds.width, bounds.height));
       AdjustForScrollBars(screenBounds, scrollableFrame);
-      metrics.mCompositionBounds = metrics.mCompositionBounds.ForceInside(screenBounds);
+      metrics.mCompositionBounds = screenBounds;
       useWidgetBounds = true;
     }
   }
