diff --git a/gfx/layers/ipc/AsyncPanZoomController.cpp b/gfx/layers/ipc/AsyncPanZoomController.cpp
index e64f803..23e25f7 100644
--- a/gfx/layers/ipc/AsyncPanZoomController.cpp
+++ b/gfx/layers/ipc/AsyncPanZoomController.cpp
@@ -1726,6 +1726,9 @@ void AsyncPanZoomController::NotifyLayersUpdated(const FrameMetrics& aLayerMetri
     if (mFrameMetrics.mViewport.width != aLayerMetrics.mViewport.width ||
         mFrameMetrics.mViewport.height != aLayerMetrics.mViewport.height) {
       needContentRepaint = true;
+      if (mFrameMetrics.mViewport.width != aLayerMetrics.mViewport.width) {
+        mFrameMetrics.SetZoom(aLayerMetrics.GetZoom());
+      }
     }
     mFrameMetrics.mViewport = aLayerMetrics.mViewport;
   }
