diff --git a/embedding/embedlite/tests/content/embedScrollStyles.css b/embedding/embedlite/tests/content/embedScrollStyles.css
index 81353bc..5bc5c5d 100644
--- a/embedding/embedlite/tests/content/embedScrollStyles.css
+++ b/embedding/embedlite/tests/content/embedScrollStyles.css
@@ -19,6 +19,8 @@ html xul|scrollbar {
   z-index: 2147483647;
   pointer-events: none;
   opacity: 1;
+  /* make frame scrollbar invisible until bug 977015 is fixed */
+  display: none;
 }
 
 xul|scrollbar[orient="vertical"] {
diff --git a/layout/generic/nsGfxScrollFrame.cpp b/layout/generic/nsGfxScrollFrame.cpp
index 7f528be..0b1de6e 100644
--- a/layout/generic/nsGfxScrollFrame.cpp
+++ b/layout/generic/nsGfxScrollFrame.cpp
@@ -2422,8 +2422,8 @@ ScrollFrameHelper::BuildDisplayList(nsDisplayListBuilder*   aBuilder,
     // The check for scroll bars was added in bug 825692 to prevent layerization
     // of text inputs for performance reasons. However, if a text input is
     // focused we want to layerize it so we can async scroll it (bug 946408).
-    bool wantLayerV = isVScrollable && (mVScrollbarBox || isFocused);
-    bool wantLayerH = isHScrollable && (mHScrollbarBox || isFocused);
+    bool wantLayerV = isVScrollable;
+    bool wantLayerH = isHScrollable;
     // TODO Turn this on for inprocess OMTC on all platforms
     bool wantSubAPZC = (XRE_GetProcessType() == GeckoProcessType_Content);
 #ifdef XP_WIN
