From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Thu, 19 Mar 2020 13:56:02 +0200
Subject: [PATCH] [sailfishos][compositor] Make it possible to extend
 CompositorBridgeParent

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
---
 gfx/layers/ipc/CompositorBridgeParent.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gfx/layers/ipc/CompositorBridgeParent.h b/gfx/layers/ipc/CompositorBridgeParent.h
index 3547c1af8c32..25c8db6daea8 100644
--- a/gfx/layers/ipc/CompositorBridgeParent.h
+++ b/gfx/layers/ipc/CompositorBridgeParent.h
@@ -198,7 +198,7 @@ class CompositorBridgeParentBase : public PCompositorBridgeParent,
   RefPtr<CompositorManagerParent> mCompositorManager;
 };
 
-class CompositorBridgeParent final : public CompositorBridgeParentBase,
+class CompositorBridgeParent : public CompositorBridgeParentBase,
                                      public CompositorController,
                                      public CompositorVsyncSchedulerOwner {
   friend class CompositorThreadHolder;
-- 
2.26.2

